File: Packages/macOS/CmuxTerminalBackend/Sources/CmuxTerminalBackend/CanonicalScreen.swift
/// One canonical screen containing a pane layout and its panes.
public struct CanonicalScreen: Codable, Equatable, Sendable {
    /// The daemon-local numeric screen identifier.
    public let id: UInt64

    /// The stable screen identifier used across daemon restarts.
    public let uuid: ScreenID

    /// The optional canonical screen name.
    public let name: String?

    /// The split tree that references every pane exactly once.
    public let layout: CanonicalLayout

    /// The canonical panes contained by this screen.
    public let panes: [CanonicalPane]

    /// Creates a canonical screen.
    ///
    /// - Parameters:
    ///   - id: The daemon-local numeric screen identifier.
    ///   - uuid: The stable screen identifier.
    ///   - name: The optional canonical screen name.
    ///   - layout: The split tree for the screen's panes.
    ///   - panes: The canonical panes contained by the screen.
    public init(
        id: UInt64,
        uuid: ScreenID,
        name: String?,
        layout: CanonicalLayout,
        panes: [CanonicalPane]
    ) {
        self.id = id
        self.uuid = uuid
        self.name = name
        self.layout = layout
        self.panes = panes
    }
}

File: cmux-tui/bindings/java/src/com/cmux/CanonicalScreen.java
package com.cmux;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public record CanonicalScreen(
    long id,
    UUID uuid,
    String name,
    CanonicalLayout layout,
    List<CanonicalPane> panes
) {
    @SuppressWarnings("unchecked")
    static CanonicalScreen from(Map<String, Object> map) {
        List<CanonicalPane> panes = new ArrayList<>();
        for (Object item : TopologyWire.list(map.get("panes"), "canonical panes")) {
            panes.add(CanonicalPane.from((Map<String, Object>) item));
        }
        return new CanonicalScreen(
            TopologyWire.uint(map.get("id")),
            TopologyWire.uuid(map.get("uuid")),
            TopologyWire.nullableString(map.get("name")),
            CanonicalLayout.from(map.get("layout")),
            List.copyOf(panes)
        );
    }
}
